plugins/org.locationtech.udig.project.ui/src/org/locationtech/udig/project/ui/internal/tool/display/ModalToolCategory.java

```diff
@@ -84,61 +84,57 @@ protected IHandler getHandler() {
     }
 
     /**
-     * Returns  a  current UI contribution item for this category of modal tools.
+     * Returns a current UI contribution item for this category of modal tools.
      *
      * @return
      */
     public AbstractToolbarContributionItem getContribution() {
         return contribution;
     }
+
     /**
-     * Hooks the provided container up to this cateogry; so that it can update
-     * in response to key presses
+     * Hooks the provided container up to this category; so that it can update in response to key
+     * presses
      *
      * @param container
      */
-    public void hook( PaletteContainer container ){
+    public void hook(PaletteContainer container) {
         this.container = container;
     }
 
-    public void container( PaletteContainer container ){
+    public void container(PaletteContainer container) {
         this.container = container;
     }
 
-    public void contribute( IToolBarManager manager ) {
-        CurrentContributionItem old = contribution;
-        contribution = new CurrentModalToolContribution();
-        manager.add(contribution);
-        //Add CurrentModalToolContribution as a contribution to each item of themodal category
-        for( ModalItem item : this ) {
-            ToolProxy tool = (ToolProxy) item;
-            if (tool.getType() == ToolProxy.MODAL) {
-                if(old != null){
-                    /*
-                     * Vitalis:
-                     * Just to clean old contributions being collected
-                     * during maps opening/closing.
-                     */
-                    tool.removeContribution(old);
+    public void contribute(IToolBarManager manager) {
+        if (contribution == null) {
+            contribution = new CurrentModalToolContribution(getId());
+            for (ModalItem item : this) {
+                ToolProxy tool = (ToolProxy) item;
+                if (tool.getType() == ToolProxy.MODAL) {
+                    tool.addContribution(contribution);
```

Comment:
Am I correct that addCorntribution ignores is if it has been added before. if so its not required to remove old first?

However, I was trying hard to reproduce the given workflow (edit layer, close map, re-open it, etc) but I couldn't. However it might be a problem if using a View rather than an editor. However, tested this fix and it doesn't break anything with MapEditorWithPalette - so good to go for me